Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,183 @@
   (with-slots (properties) port
     (setf (getf properties indicator) value)))
 
+;;; This function determines the sheet to which the pointer event
+;;; should be delivered. The right thing is not obvious:
+;;;
+;;; - we may assume that event-sheet is set correctly by port
+;;; - we may find the innermost sheet's child and deliver to it
+;;; - we may deliver event to sheet's graft and let it dispatch
+;;;
+;;; Third option would require a default handle-event method to call
+;;; handle-event on its child under the cursor. For now we implement
+;;; the second option with the innermost child. In general case we
+;;; need z-ordering for both strategies. -- jd 2019-08-21
+(defun compute-pointer-event-sheet (event &aux (sheet (event-sheet event)))
+  (labels ((gpes-rec (sheet x y)
+             (if-let ((child (child-containing-position sheet x y)))
+               (multiple-value-call #'gpes-rec
+                 child (untransform-position (sheet-transformation child) x y))
+               sheet)))
+    (get-pointer-position (sheet event)
+      (gpes-rec (event-sheet event) x y))))
+
+;;; Function is responsible for making a copy of an immutable event
+;;; and adjusting its coordinates to be in the target-sheet
+;;; coordinates. Optionally it may change event's class.
+(defun dispatch-event-copy (target-sheet event &optional new-class
+                            &aux (sheet (event-sheet event)))
+  (if (and (eql target-sheet sheet)
+           (or (null new-class)
+               (eql new-class (class-of event))))
+      (dispatch-event sheet event)
+      (let ((new-event (shallow-copy-object event)))
+        (get-pointer-position (target-sheet new-event)
+          (setf (slot-value new-event 'x) x
+                (slot-value new-event 'y) y
+                (slot-value new-event 'sheet) target-sheet))
+        (when new-class
+          (change-class new-event new-class))
```

Comment:
it is not guaranteed that this other class will have the same slots as the original one. change-class reflects the purpose much better (imho).

Reply:
I don't agree with the purpose part. `change-class` has to "diff" the slots of the old and the new class, introspecting the slot list a second time, not to mention potential type checks and initarg checks -- this is very expensive. Why not allocate an instance of the required class and copy only the relevant slots right from the start?

This would be enough:
```lisp
(defun shallow-copy-object (original &optional (new-class (class-of original)))
  (let ((copy (allocate-instance new-class)))
    (dolist (slot (mapcar #'c2mop:slot-definition-name
                          (c2mop:class-slots new-class)))
      (when (and (slot-exists-p original slot)
                 (slot-boundp original slot))
        (setf (slot-value copy slot)
              (slot-value original slot))))
    copy))
```

Reply:
fixed (shallow-copy-object was modified in a separate commit before the distribute-event changes)